val.go

```diff
@@ -292,8 +292,8 @@ func (v *RawVal) Observe(val float64) {
 // Stats implements the StatSource interface.
 func (v *RawVal) Stats(cb func(key SeriesKey, field string, val float64)) {
 	v.mtx.Lock()
+	defer v.mtx.Unlock()
 	cb(v.key, "recent", v.value)
-	v.mtx.Unlock()
 	for _, s := range v.stats {
 		field, value := s()
 		cb(v.key, field, value)
```

Comment:
monkit is careful in all of the other stat sources to not call into the callback while holding the mutex to avoid potential deadlocks and not limit concurrency for functions that are trying to mutate the value while stat collection is happening.

Reply:
I understand that goal, but I think this situation is different. Sum/Count are very compact functions, and I think it's better to lock only once, not multiple times.

What is your suggestion?

Reply:
so it's not the sum/count callbacks, it's the stats callback. you also don't need to lock multiple times. i'm thinking of something like:

```go
type datum struct { field string; value float64 }
data := make([]datum, len(v.stats))
v.mtx.Lock()
for i, s := range v.stats {
    field, value := s()
    data[i] = datum{field, value}
}
recent := v.value
v.mtx.Unlock()

cb(v.key, "recent", v.value)
for _, d := range data {
    cb(v.key, d.field, d.value)
}
```

then to avoid the allocation in the common case of a small amount of aggregators, you can do

```go
var data []datum
if len(v.stats) < 8 {
    var buf [8]datum
    data = buf[:len(v.stats)]
} else {
    data = make([]datum, len(v.stats))
}
```